[SYCL] Introduce extension proposal for asynchornous memory management (#14800)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

This extension also depends on the following other SYCL extensions:

* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
  sycl_ext_oneapi_properties]
* link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions]
* link:../supported/sycl_ext_oneapi_default_context.asciidoc[
  sycl_ext_oneapi_default_context]

== Status

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Backend support status

This extension is not yet implemented in {dpcpp}.

== Overview

This extension introduces an interface for allocating and freeing USM memory
asynchronously, as well as introducing a USM memory pool from which this memory
is allocated and freed from.

> The asynchornous memory allocation and free are exposed via new free functions

Reviewer: ```suggestion
The asynchronous memory allocation and free are exposed via new free functions
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

taking a queue, which enqueue an asynchronous malloc or free command and return
immediately. The functions can take a memory pool object in order to allocate
from that memory pool specifically, however, if no memory pool object is
provided there is a default memory pool which will be used.

The memory pool introduced is a dynamic memory pool, as such no memory is
immediately allocated on construction, instead memory is allocated by the OS
for the pool when allocations are made and are freed back to the OS when the

> queue which enqueued the malloc and free commands is synchronized with.

Reviewer: Any reason we have to be specific about the OS here that does the allocation, specifically when? or can we just say runtime? Ultimately yes the allocation will be done at the OS level, but the way it's phrased could prevent optimizations like caching on runtime level. This could just be implementation specific.

Author: Yeah that's a good point, I think we should rephrase this to say allocated by the SYCL RT.

> The immediate benefit of using the asynchornous malloc and free functions, as

Reviewer: ```suggestion
The immediate benefit of using the asynchronous malloc and free functions, as
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

opposed to the existing synchronous malloc and free functions, is that they no
longer block the calling thread. So applications which wish to allocate

> temporary USM memory but not carry around the pointer no longer have to

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

synchronize with the execution of kernels accessing that memory and call the
free function before continuing.

A further benefit comes from the use of a memory pool for allocations, as
malloc and free commands are asynchronous, the runtime can opportunistically
re-use memory allocations from one kernel to another, without any intermediate
freeing and re-allocation. This benefit can be further extended beyond queue

> synchronization by specify a threshold, which will instruct the runtime to (if

Reviewer 2: ```suggestion
synchronization by specifying a threshold, which will instruct the runtime to (if
```

possible) maintain a certain size of memory and not release that back to the OS
even when the queue is synchronized with.

There are also other properties which can be used when constructing a memory
pool object to control other aspects of how the memory is allocated.

== Integration with other extensions

It is a priority for this extension to integrate with the

> link:../supported/sycl_ext_oneapi_graph.asciidoc[

Reviewer: ```suggestion
link:../experimental/sycl_ext_oneapi_graph.asciidoc[
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

> sycl_ext_oneapi_graph] extension. The asynchornous commands introduces in this

Reviewer 3: asynchronous

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

extension are intended to be compatible with the SYCL Graph record and replay
mode, however, the specifics of this will be defined in that extension.

It is a priority for this extension to integrate with the

> link:../supported/sycl_ext_oneapi_bindless_images.asciidoc[

Reviewer: ```suggestion
link:../experimental/sycl_ext_oneapi_bindless_images.asciidoc[
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

sycl_ext_oneapi_bindless_images] extension. It is the intention that the memory
allocations from the memory pool in this extension be used for the bindless
images, however, this will need to be investigated further and so the details of
this are not included in this version of the extension.

== Examples

There are various ways in which this extension can be used but a typical usage
of the memory pool and the asynchronous malloc and free commands is described
below. In this example an explicit memory pool is created and this is used to

> share memory allocated from the OS between multiple asynchornous malloc

Reviewer 3: asynchronous

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

commands.

[source,c++]
----
int main(int argc, char *argv[])
{
  queue q(property::queue::in_order);

  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
    q.get_device(), usm::alloc::device);
  
  {
    // memory pool allocates memory from the OS
    void *temp = async_malloc_from_pool(q, 1024, memPool);

    // memory allocation is used for first kernel
    parallel_for(q, range{1024}, [=](id<1> idx) {
      do_something(idx, temp);
    });

    // memory is available to be used by another allocation
    async_free(q, temp);
  }

  {
    // memory pool re-uses previously allocated memory
    void *temp = async_malloc_from_pool(q, 1024, memPool);

    // memory allocation is used for second kernel
    parallel_for(q, range{1024}, [=](id<1> idx) {
      do_something_else(idx, temp);
    });

    // memory is available to be used by another allocation
    async_free(q, temp);
  }

  // memory pool releases memory back to the OS

>   q.wait();

Reviewer 2: the memory pool is created with q.context and q.device, it is not bound with queue itself. What will happen if multiple queues are used as below code.

```
queue q1(inorder), q2(inorder);
memory_pool memPool(q1.get_context(), q1.get_device(), usm::alloc::device);

void *temp1 = async_malloc_from_pool(q1, 1024, memPool);
async_free(q1, temp1);

void *temp2 = async_malloc_from_pool(q2, 1024, memPool);
async_free(q2, temp2);

q1.wait();
```

Author: This is a good question, this isn't specified in the extension right now. I would say that since the memory pool is tied to a context/device pair, `async_malloc`/`async_malloc_from_pool` should require that the device associated with the queue is one of the devices associated with the memory pool.

Author: I've clarified the wording of `async_malloc` and `async_malloc_from_pool` to state that the queue and the memory pool must be associated with the same context and device.

[source,c++]
----
int main(int argc, char *argv[])
{
  queue q;

  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
    q.get_device(), usm::alloc::device);
  
  {
    void *temp = null;

    // memory pool allocates memory from the OS
    auto e1 = q.submit_with_event([&](handler &cgh) {
      temp = async_malloc_from_pool(q, 1024, memPool);
    });

    // memory allocation is used for first kernel
    auto e2 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e1);
      parallel_for(q, range{1024}, [=](id<1> idx) {
        do_something(idx, temp);
      });
    });

    // memory is available to be used by another allocation
    auto e3 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e2);
      async_free(q, temp);
    });
  }

  {
    void *temp = null;

    // memory pool re-uses previously allocated memory
    auto e1 = q.submit_with_event([&](handler &cgh) {
      temp = async_malloc_from_pool(q, 1024, memPool);
    });

    // memory allocation is used for second kernel
    auto e2 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e1);
      parallel_for(q, range{1024}, [=](id<1> idx) {

>         do_something_else(idx, temp);

Reviewer: Not sure if that works as expected with a just plain pointer. `temp` is captured by value by the kernel function, but there is no guarantee that the final value is available. Because the preceding allocation command is supposed to be executed asynchronously.

Reviewer 2: I also have the concern, the 'first 3 submit_with_event' and the 'last 3 submit_with_event' are independent, if we still have hardware resource available when do_something is executed, it is possible that do_something_else is scheduled to execution concurrently.

Reviewer 2: looks that 74d6487 does not change meaningful code.

> there should be dependency between the first async_free and the second async_malloc_from_pool
why there's dependency between them?

Author: Good catch I missed the `depends_on` in the fourth command group, fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

      do_something_else(idx, temp);
    });

    // memory is available to be used by another allocation
    async_free(q, temp);
  }

  // again memory pool does not release memory back to the OS
  q.wait();
}
----

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
table below.  Applications can test for the existence of this macro to determine
if the implementation supports this feature, or applications can test the
macro's value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Memory pool

This extension introduces the `memory_pool` class, which provides a handle to a
memory pool owned by the SYCL runtime or a specific backend, and adheres to the
SYCL common reference semantics.

Memory pools have the following properties:

> * An allocation max size (in bytes) is used to manage to total amount of memory

Reviewer 2: ```suggestion
* An allocation max size (in bytes) is used to manage the total amount of memory
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

  which can be allocated in the memory pool. If the max size is exceeded an
  error is triggered. The default max size is implementation-defined.
* A deallocation threshold (in bytes) is used to determine how much memory the
  SYCL runtime should aim to maintain in the memory pool, without releasing back
  to the OS. The default deallocation threshold is zero.
* No memory is immediately allocated on construction, instead memory is
  allocated from the OS for the pool when requested via the asynchronous malloc
  and free functions.
* Memory is freed back to the OS when the SYCL queue which enqueued the
  respective allocations and frees is synchronized with, i.e. when `queue::wait`
  or `queue::wait_and_throw` is called, unless the current total memory
  allocated to the memory pool under the deallocation threshold, in which case

>   the memory may be retained by the pool, though this can vary dependening on

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

  implementation defined parameters.
* Memory is allocated as USM memory, in one of the USM memory allocation kinds
  enumerated in `usm::alloc`, this is specified on construction of the
  `memory_pool` object.
* They are associated with a specific context and one or more device(s),
  depending on the allocation kind, this is specified on construction of the
  `memory_pool` object.

Memory pools are intended to be used for both in-order and out-of-order SYCL
queues.

[source,c++]
----
namespace ext::oneapi::experimental {

class memory_pool {

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, Properties props = {});

  template <typename Properties = empty_properties_t>

>   memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});

Author: That's a good idea, I've added an overload for this in https://github.com/intel/llvm/pull/14800/commits/48493ace0babceba6b65fe7399b43284c10681db.

pool to be destroyed after all remaining allocations have been freed, and
returns immediately without waiting.

[source, c++]
----
context get_context() const;
----

_Returns_: The SYCL context associated with the memory pool.

[source, c++]
----
std::vector get_devices() const;
----

_Returns_: The SYCL device(s) associated with the memory pool.

[source, c++]
----
usm::alloc get_alloc_kind() const;
----

_Returns_: The memory allocation kind of the memory pool.

[source, c++]
----
size_t get_max_size() const;
----

_Returns_: The maximum size of the memory pool.

[source, c++]
----
size_t get_threshold() const;
----

_Returns_: The deallocation threshold of the memory pool.

[source, c++]
----
void set_new_threshold(size_t newThreshold);
----

_Effects_: Sets the deallocation threshold of the memory pool if the value of
`newThreshold` is larger than the current threshold.

_Throws_: An exception with the `errc::invalid` error code if the value of
`newThreshold` is lower than the current threshold.

=== Memory pool properties

A memory pool can be constructed with a number of properties which can change

> certain behaviours, these can be specified when constructing a `memory_pool`

Reviewer: ```suggestion
certain behaviors, these can be specified when constructing a `memory_pool`
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

};

inline constexpr initial_threshold_key::value_t initial_threshold;
inline constexpr maximum_size_key::value_t maximum_size;
inline constexpr read_only_key::value_t read_only;
inline constexpr zero_init_key::value_t zero_init;

}  // ext::oneapi::experimental
----

|===
|Property|Description

|`initial_threshold`
|The `initial_threshold` property specifies the initial deallocation threshold
 value for the memory pool. If this property is not used the default value is
 zero, and this can be increased after the memory pool is created by calling
 `memory_pool::set_new_threshold`.

|`maximum_size`
|The `maximum_size` property specifies the maximum size of the memory pool,
 after which any allocation will result in an exception. If the value specified
 is larger than the implementation can support an exception with the
 `errc::memory_allocation` error code is thrown. If this property is not used
 the default value is implementation-defined.

|`read_only`
|The `read_only` property is a performance hint which asserts that all memory
 allocations from the memory pool will only ever be read from, this can be used
 by the SYCL runtime to optimize for performance.

|`zero_init`

> |The `zero_init` property adds the rquirement that all memory allocated to the

Reviewer 4: ```suggestion
|The `zero_init` property adds the requirement that all memory allocated to the
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

 memory as it is allocated from the OS to the memory pool will be initialised to
 zero. Note there is no guarantee that the memory allocation be re-initialized
 to zero when it is re-allocated from the pool, so users must re-initialize

>  memory to zero if they wish for later allocations to have this behaviour.

Reviewer: ```suggestion
 memory to zero if they wish for later allocations to have this behavior.
```

|===

=== Default memory pools

As well as being able to construct a memory pool explicitly, this extension
introduces a default memory pool per device for each SYCL context and device
pair for device allocations and a default memory pool per context for host
allocations.

New member functions are added to the `context` class to retrieve the default
memory pool as a copy of the `memory_pool` object. This can be modified and have
those modifications reflected as it conforms to the SYCL common reference
semantics.

[source,c++]
----
class context {

  memory_pool context::ext_oneapi_get_default_memory_pool();

  memory_pool context::ext_oneapi_get_default_memory_pool(device dev);

}; // context
----

[source, c++]
----

> memory_pool context::ext_oneapi_get_default_memory_pool();

Reviewer 4: ```suggestion
memory_pool context::ext_oneapi_get_default_memory_pool() const; 
```
Not sure if these can be `const` if they are query methods

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

----

_Returns_: The default memory pool associated with the context for allocating
with the allocation kind `usm::host`.

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool(device dev);
----

_Returns_: The default memory pool associated with the context and `dev` for
allocating with the allocation kind `usm::device`.

=== Asynchronous malloc & free

This extension introduces a series of new enqueue functions for enqueueing

> asynchornous malloc and free commands which operate with the memory pools also

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

introduced in this extension.

All enqueue functions introduced have overloads which take a SYCL `queue` and a
SYCL `handler`. None of enqueue functions return a SYCL `event` direction, as
this extension is in line with the
link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
calling `submit_with_event` and the `handler` overloads of these enqueue
functions.

[source,c++]
----
namespace ext::oneapi::experimental {

void *async_malloc(queue q, size_t size);

void *async_malloc(handler h, size_t size);

void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);

void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);

void async_free(queue q, void *ptr);

void async_free(handler h, void *ptr);

}  // ext::oneapi::experimental
----

[source, c++]
----
void *async_malloc(queue q, size_t size);

void *async_malloc(handler h, size_t size);

void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);

void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
----

_Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`

> which will asynchornously allocate memory of size `size` in bytes, allocating

Reviewer 4: ```suggestion
which will asynchronously allocate memory of size `size` in bytes, allocating
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

from the memory pool `pool` if provided, otherwise allocation from the default
memory pool associated with the SYCL context and device associated with `q` or
`h`. Memory is first allocated from the memory pool if possible, otherwise
memory is allocated from the OS to the memory pool to provide enough memory in
the memory pool for the allocation.

_Returns_: A pointer to the address of a memory reservation.

_Throws_: An exception with the `errc::invalid` error code if `size` is zero.
An exception with the `errc::memory_allocation` error code if the allocation
brings the memory pool over the maximum size. This error can be thrown

> asynchornously.

Reviewer: ```suggestion
asynchronously.
```

[_Note:_ Accessing the memory at the address of the pointer returned by

> asynchornous malloc functions before the command has completed execution is

Reviewer 2: what does 'the command' represent?  the asynchronous malloc functions?

Author: So the command is a SYCL RT command that executes the asynchronous memory allocation from the memory pool, so until this command is complete is undefined behaviour to access the memory at that address.

Reviewer 2: does this specific for out-of-order sycl queue?  If it is in-order sycl queue, I think there's no issue to first allocate memory async and then just access that memory address in a kernel.

Author: So the pointer can be passed to a kernel that is enqueued, but users must ensure that either the commands are either enqueued to the same in-order queue in which case, as you mentioned this will always be correct as the queue scheduling semantics guarantees correctness, or if enqueued to separate queues or an out-of-order queue the necessary dependencies are established.

> asynchornous malloc functions before the command has completed execution is
> undefined behaviour. _{endnote}_]

Reviewer: ```suggestion
asynchronous malloc functions before the command has completed execution is
undefined behavior. _{endnote}_]
```

[source, c++]
----
void async_free(queue q, void *ptr);

void async_free(handler h, void *ptr);
----

_Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`

> which will asynchornously free the memory allocation at the address of `ptr`.

Reviewer 4: ```suggestion
which will asynchronously free the memory allocation at the address of `ptr`.
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

> _Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`
> which will asynchornously free the memory allocation at the address of `ptr`.
> Memory will be freed from the memory pool to be used by other asynchronous
> malloc commands which execute later, and will not free to the OS until the SYCL
> queue associated with the asycnhronous allocation command has been synchornized
> with.

Reviewer: ```suggestion
_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h`
which will asynchronously free the memory allocation at the address of `ptr`.
Memory will be freed from the memory pool to be used by other asynchronous
malloc commands which execute later, and will not free to the OS until the SYCL
queue associated with the asynchronous allocation command has been synchronized
with.
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

_Throws_: An exception with the `errc::invalid` error code if `ptr` is not the

> address of a memory allocation allocated to a memory pool.

Reviewer 2: another exception that if the `q`/`h` is different than the queue to create `ptr`？

Reviewer 2: > another exception that if the `q`/`h` is different than the queue to create `ptr`？

please ignore this comment, just see issue 2.

> [_Note:_ Accessing the memory at the address of `ptr`after the asynchronous free

Reviewer 4: ```suggestion
[_Note:_ Accessing the memory at the address of `ptr` after the asynchronous free
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

command has completed execution is undefined behaviour. _{endnote}_]

=== Memory pool lifetimes

The lifetime of memory allocated via a memory pool is as follows.

> When an asynchornous malloc command is executed it will first look to the

Reviewer 4: ```suggestion
When an asynchronous malloc command is executed it will first look to the
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

memory pool to opportunistically allocate memory already available there. The
SYCL RT may also, when possible, look at all asynchronous malloc and free
commands and their dependencies in order to more efficiently schedule memory
allocations. If the memory pool does not have the requirement memory available
to allocate for the asynchronous malloc command it will then look to allocate
from the OS to the memory pool to provide the required memory.

> When an asynchornous free command is executed it will free up memory in the 

Reviewer 4: ```suggestion
When an asynchronous free command is executed it will free up memory in the 
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

memory pool to be used for a later asynchonrous malloc command. Memory is not
freed back to the OS at this point.

When a SYCL queue which enqueued asynchronous malloc commands is synchronized

> with, i.e. when `queue::wait` oo `queue::wait_and_throw` is called, the SYCL RT\

Reviewer 4: ```suggestion
with, i.e. when `queue::wait` or `queue::wait_and_throw` is called, the SYCL RT
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

> will check the current total memory allocated after to the memory pool against

Reviewer 2: ```suggestion
will check the current total memory allocated to the memory pool against
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

> the memory pool's deallocation threshold. If the total memory allocated islarger

Reviewer 2: ```suggestion
the memory pool's deallocation threshold. If the total memory allocated is larger
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

than than the threshold then the memory will be freed back to the OS, otherwise
it will not.

[_Note:_ An implementation should maintain memory allocated to a memory pool to
the size specified by the deallocation threshold when possible, however, an
implementation is permitted to release this memory back to the OS if it is
required by another memory pool or an application in another process.
_{endnote}_]

== Implementation notes

It is expected that for L0 this extension will be implemented within the L0
adapter, by reserving allocations for the memory pool and opportunistically
re-using the memory allocated based on the command lists being enqueued to the
L0 driver.

It is expected that for CUDA this extension will be implemented by mapping onto
the CUDA stream-ordered allocator feature.

== Issues

. Should we allow mixing asynchornous and synchornous memory commands?
+
--

> *UNRESOLVED*: CUDA allows memory allocated with the asynchornous malloc command
> to be freed with the regular sycnhronous free command, should we extend this

Reviewer: ```suggestion
*UNRESOLVED*: CUDA allows memory allocated with the asynchronous malloc command
to be freed with the regular synchronous free command, should we extend this
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

capability to SYCL?
--

. Should we allow freeing memory with a different queue?
+
--
*UNRESOLVED*: Should we allow a memory allocation allocated with an asynchronous
malloc command from one queue to be freed by an asynchronous free command from
another queue?
--

. How should we treat an asynchronous malloc command with a memory pool that is
not associated with that queue?
+
--
*UNRESOLVED*: If an asynchornous malloc command is enqueued with a memory pool
that is not associated with the queue that the command is enqueued from, should
this result in an error?
--

. Should we have a default memory pool for `usm::shared`?
+
--
*UNRESOLVED*: Currently the proposed API means that there cannot be default
memory pool for allocations of allocation kind `usm::shared`, and therefore a
user must create their own explicit memory pool to do so. Is this reasonable or
should we extend the API to include a default memory pool for allocations of
allocation kind `usm::shared`?
--

. Should we allow setting a new threshold that is lower?
+
--
*UNRESOLVED*: Currently setting a new deallocation threshold is only permitted
if it increases the size of the threshold, however, we may want to also allow

> setting a new lower threshold. This would work by by not immediately freeing any

Reviewer: ```suggestion
setting a new lower threshold. This would work by not immediately freeing any
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

> memory but using this lower threshold at the next synchronisation point.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.

== Backend support status

This extension is not yet implemented in {dpcpp}.

== Overview

This extension introduces an interface for allocating and freeing USM memory
asynchronously, as well as introducing a USM memory pool from which this memory
is allocated and freed from.

The asynchronous memory allocation and free are exposed via new free functions
taking a queue, which enqueue an asynchronous malloc or free command and return
immediately. The functions can take a memory pool object in order to allocate
from that memory pool specifically, however, if no memory pool object is
provided there is a default memory pool which will be used.

The memory pool introduced is a dynamic memory pool, as such no memory is
immediately allocated on construction, instead memory is allocated by the SYCL
runtime or an existing memory allocation provided by the user, for the pool when
allocations are made and are freed back to the SYCL runtime or the memory
allocation provided by the user, when the queue which enqueued the malloc and
free commands is synchronized with.

The immediate benefit of using the asynchronous malloc and free functions, as

> opposed to the existing synchronous malloc and free functions, is that they no

Reviewer 2: ```suggestion
opposed to the existing synchronous malloc and free functions, is that the free function no
```

the second sentence in this paragraph explains that why sync free function blocks the calling thread. 

Does the sync malloc function block the calling thread? (I guess the answer is no?)

Author: That's a good point, the sync `malloc` doesn't necessarily block the calling thread to allocate the memory, and the implementation of `async_malloc` may block, so this statement is a bit misleading, I've clarified this in https://github.com/intel/llvm/pull/14800/commits/044ec4427945bcbd526ed3874e0989fe8ffa54e7.

extension are intended to be compatible with the SYCL Graph record and replay
mode, however, the specifics of this will be defined in that extension.

It is a priority for this extension to integrate with the
link:../experimental/sycl_ext_oneapi_bindless_images.asciidoc[
sycl_ext_oneapi_bindless_images] extension. It is the intention that the memory
allocations from the memory pool in this extension be used for the bindless
images, however, this will need to be investigated further and so the details of
this are not included in this version of the extension.

== Examples

There are various ways in which this extension can be used but a typical usage
of the memory pool and the asynchronous malloc and free commands is described
below. In this example an explicit memory pool is created and this is used to
share memory allocated from the SYCL runtime between multiple asynchronous
malloc commands.

[source,c++]
----
int main(int argc, char *argv[])
{
  queue q(property::queue::in_order);

  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
    q.get_device(), usm::alloc::device);
  
  {
    // memory pool allocates memory from the SYCL runtime

>     void *temp = async_malloc_from_pool(q, 1024, memPool);

Author: I've updated the examples to use `syclex` namespace aliases as you described, changed in https://github.com/intel/llvm/pull/14800/commits/a7bcd952714d62ca2663ea24a6037bd4a8d1c994.

      temp = async_malloc_from_pool(cgh, 1024, memPool);
    });

    // memory allocation is used for second kernel
    auto e5 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e4);
      parallel_for(cgh, range{1024}, [=](id<1> idx) {
        do_something_else(idx, temp);
      });
    });

    // memory is available to be used by another allocation
    q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e5);
      async_free(cgh, temp);
    });
  }

  // memory pool releases memory back to the SYCL runtime
  q.wait();
}
----

Another example of memory pool usage is described in the example below. In this
example rather than creating an explicit memory pool the default memory pool is
being used instead. There is also additional queue synchronization between the
commands enqueued which would ordinarily lead to memory being released back to
the SYCL runtime, however, the allocation threshold for the memory pool is
extended so the memory pool maintains the allocations and therefore still
provide the benefit of re-allocating memory from the memory pool.

[source,c++]
----
int main(int argc, char *argv[])
{
  queue q(property::queue::in_order);

>   ext::oneapi::experimental::memory_pool memPool

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/044ec4427945bcbd526ed3874e0989fe8ffa54e7.

macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
table below.  Applications can test for the existence of this macro to determine
if the implementation supports this feature, or applications can test the
macro's value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Memory pool

This extension introduces the `memory_pool` class, which provides a handle to a
memory pool owned by the SYCL runtime or a specific backend, and adheres to the
SYCL common reference semantics.

Memory pools have the following properties:

* A memory pool can allocate memory from two possible sources; either the SYCL
  runtime or an existing USM memory allocation provided by the user. The default
  source is the SYCL runtime.
* A maximum allocation size (in bytes) is used to manage the total amount of
  memory which can be allocated in the memory pool. If the maximum size is
  exceeded an error is thrown. The default maximum size is

>   implementation-defined.

Author: That's a good point, I've added a clarification that it can't be zero in https://github.com/intel/llvm/pull/14800/commits/f53598e0d8f33105d00fad526454dc08ab56b762.

* A deallocation threshold (in bytes) is used to determine how much memory the
  SYCL runtime should aim to maintain in the memory pool, without releasing back

>   to the source. The default deallocation threshold is zero.

Reviewer 5: `source` is an overloaded word. Can you clarify the wording here a bit more, please.

Reviewer 5: I think in general I don't understand what this property is doing. `Used to determine`, `should`, ... is quite vague phrasing IMO. It would be good if you supplemented this with an example.

Author: Use of the term source has been removed in https://github.com/intel/llvm/pull/14800/commits/044ec4427945bcbd526ed3874e0989fe8ffa54e7.

  or `queue::wait_and_throw` is called, unless the current total memory
  allocated to the memory pool under the deallocation threshold, in which case
  the memory may be retained by the pool, though this can vary depending on
  implementation defined parameters.
* Memory is allocated as USM memory, in one of the USM memory allocation kinds
  enumerated in `usm::alloc`, this is specified on construction of the
  `memory_pool` object.
* They are associated with a specific context and one or more device(s),
  depending on the allocation kind, this is specified on construction of the
  `memory_pool` object.

Memory pools are intended to be used for both in-order and out-of-order SYCL
queues.

[source,c++]
----
namespace ext::oneapi::experimental {

class memory_pool {

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});

  ~memory_pool();

  context get_context() const;

  std::vector<device> get_devices() const;

  usm::alloc get_alloc_kind() const;

>   size_t get_max_size() const;

Author: I've clarified the wording for `get_max_size` to describe the maximum size of the memory pool. It's a good point about the maximum allocation size, do you think we should have another query for this?

>   size_t get_max_size() const;

Reviewer 6: > It's a good point about the maximum allocation size, do you think we should have another query for this?

I guess it depends on what we would like to achieve and which implementation we envision for this.

If this memory pool is intended simply as a wrapper on top of regular `sycl::malloc` (and its variants), then we could simply say that `async_malloc_from_pool` has the same limitations as `sycl::malloc` on top of any of its own limitations.

If we envision that an implementation could be tricker (for example, re-use virtual memory mechanism), then we should probably add corresponding query.

The first option is easier to implement and it is probably better simply from the consistency point of view: why a "scheduled" (i.e. async) `malloc` from a pool would have less restrictions than regular sync `malloc`?

>   size_t get_max_size() const;

Author: Sorry I missed this comment earlier. So I think we want to allow both of these implementations, a memory pool could be simply a wrapper around regular malloc/free or it could a more complex implementation involving virtual memory. In which case I think it would be safer to assume the same limitations as `sycl::malloc`.

  size_t get_threshold() const;

  void set_new_threshold(size_t newThreshold);

}; // memory_pool

}  // ext::oneapi::experimental
----

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(context ctx, Properties props = {});
----

_Effects_: Constructs a memory pool associated with `ctx` and all SYCL devices
associated with it, with the allocation kind `usm::alloc::host` and applying any
properties in `props`.

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
----

_Effects_: Constructs a memory pool associated with `ctx` and `dev`, with the
allocation kind `kind` and applying any properties in `props`.

_Throws_: An exception with the `errc::invalid` error code if `kind` is

> `usm::alloc::host`.

Author: There's another constructor `memory_pool(context ctx, Properties props = {})` which creates a memory pool for host allocations.

I've added a clarification that kind cannot be `alloc::unknown` in https://github.com/intel/llvm/pull/14800/commits/f53598e0d8f33105d00fad526454dc08ab56b762.

[source, c++]
----
~memory_pool();
----

_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
pool to be destroyed after all remaining allocations have been freed, and
returns immediately without waiting.

[source, c++]
----
context get_context() const;
----

_Returns_: The SYCL context associated with the memory pool.

[source, c++]
----
std::vector<device> get_devices() const;
----

_Returns_: The SYCL device(s) associated with the memory pool.

[source, c++]
----
usm::alloc get_alloc_kind() const;
----

_Returns_: The memory allocation kind of the memory pool.

[source, c++]
----
size_t get_max_size() const;
----

_Returns_: The maximum size of the memory pool.

[source, c++]
----
size_t get_threshold() const;
----

_Returns_: The deallocation threshold of the memory pool.

[source, c++]
----
void set_new_threshold(size_t newThreshold);
----

_Effects_: Sets the deallocation threshold of the memory pool if the value of
`newThreshold` is larger than the current threshold.

_Throws_: An exception with the `errc::invalid` error code if the value of
`newThreshold` is lower than the current threshold or larger than the maximum

> allocation size.

Author: So the idea of the maximum size is a size which the user can specify as a limit which the memory pool will not exceed, rather than a guarantee that this amount of memory can or will be allocated by the pool.

=== Memory pool properties

A memory pool can be constructed with a number of properties which can change
certain behaviors, these can be specified when constructing a `memory_pool`
object.

[source,c++]
----
namespace ext::oneapi::experimental {

struct initial_threshold_key {
  using value_t = property_value<initial_threshold_key>;

  initial_threshold_key(size_t initialThreshold);
};

struct maximum_size_key {
  using value_t = property_value<maximum_size_key>;

  initial_threshold_key(size_t maxSize);
};

struct read_only_key {
  using value_t = property_value<read_only_key>;

  read_only_key(bool readOnly);
};

struct zero_init_key {
  using value_t = property_value<zero_init_key>;

  read_only_key(bool zeroInit);
};

struct use_existing_memory_key {
  using value_t = property_value<use_existing_memory_key>;

>   use_existing_memory_key(void *ptr, size_t size);

Reviewer 5: What happens if the `memory_pool` is associated with a multi device context, but the pool is backed by a device allocation for a single device? Relevant mostly to HIP/CUDA backends.

Author: So when using a user provided  USM allocation in a memory_pool (which is now a different constructor rather than being a property), the device and the allocation kind is inferred from the pointer (via `get_pointer_device` and `get_pointer_type`), so in the case of host memory it is allocated for all devices in the context but for device or shared memory it is only allocated for the specific device the pointer has been allocated for.

};

inline constexpr initial_threshold_key::value_t initial_threshold;
inline constexpr maximum_size_key::value_t maximum_size;
inline constexpr read_only_key::value_t read_only;
inline constexpr zero_init_key::value_t zero_init;
inline constexpr use_existing_memory_key::value_t use_existing_memory;

}  // ext::oneapi::experimental
----

|===
|Property|Description

|`initial_threshold`
|The `initial_threshold` property specifies the initial deallocation threshold
 value for the memory pool. If this property is not used the default value is
 zero, and this can be increased after the memory pool is created by calling
 `memory_pool::set_new_threshold`.

|`maximum_size`

> |The `maximum_size` property specifies the maximum size of the memory pool,

Author: So this is the maximum size that the memory pool is allowed to become, an error will be thrown if the pool would attempt to allocate more memory than the maximum.

> |The `maximum_size` property specifies the maximum size of the memory pool,

Reviewer 6: > So this is the maximum size that the memory pool is allowed to become, an error will be thrown if the pool would attempt to allocate more memory than the maximum.

What is a use-case for that? So that users could distribute the memory to different pools and have a built-in mechanism of ensuring correctness of user assumptions about memory consumption by each pool?

> |The `maximum_size` property specifies the maximum size of the memory pool,

Author: That's right, the idea is that a user can specify the maximum size as a way of limiting the size which the memory pool can grow to, to assert that an application or library will not use more than a certain amount of memory.

 after which any allocation will result in an exception. If the value specified
 is larger than the implementation can support an exception with the
 `errc::memory_allocation` error code is thrown. If this property is not used
 the default value is implementation-defined.

|`read_only`

> |The `read_only` property is a performance hint which asserts that all memory
>  allocations from the memory pool will only ever be read from within SYCL kernel
>  functions, this can be used by the SYCL runtime to optimize for performance.

Reviewer 7: ```suggestion
|The `read_only` property is an assertion by the application that allocations
 from this pool will never be written from any kernel and will never be a
 destination of a memory copy operation that is submitted to a device.
 Implementations may be able to perform certain optimizations when a memory
 pool is created with this property.           
```

I assume the part about copy operations is what we intend?

Author: It was intentional that allocations from a memory pool with the `read_only` property would be read only in the kernel but could still be modified by the host, including copying to the memory location via a memcpy. The intended optimisations are only for the kernel size.

Reviewer 7: I suggest making this more clear with a non-normative note:

> The `read_only` property is an assertion by the application that allocations from this pool will never be written from any kernel.  Implementations may be able to perform certain optimizations when a memory pool is created with this property.
>
> [Note: A memory pool created with the `read_only` property may still be written from the host and may still be the destination of a memory copy operation that is submitted to a device. -- end note]

(Note that I intentionally reworded to avoid "only ever be read from SYCL kernel functions" because I think that phrase is unclear.)

|`zero_init`
|The `zero_init` property adds the requirement that all memory allocated to the
 memory as it is allocated from the source to the memory pool will be
 initialised to zero. Note there is no guarantee that the memory allocation be
 re-initialized to zero when it is re-allocated from the pool, so users must
 re-initialize memory to zero if they wish for later allocations to have this
 behavior.

|`use_existing_memory`
|The `use_existing_memory` property adds the requirement that the memory pool
 will use an existing USM memory allocation provided by the user instead of
 allocating from the SYCL runtime. This property takes a pointer to a valid USM
 memory allocation of the same allocation kind as the memory pool is initialized
 with and the size of that memory allocation. Using this property will

>  implicitly set the `maximum_size` and `initial_threshold` property to be that

Reviewer 5: This means that all USM allocations will need to be tracked in SYCL/UR (so that we can associate a ptr with the size of the allocation).

Also, why does this mean that the `initial_threshold` property will also be set?

Reviewer 5: Maybe good to also mention that use of a user provided ptr outside using the mem pool during the lifetime of the mem pool results in undef behaviour.

Author: So it shouldn't be necessary for the SYCL RT or UR to track the size of allocation for a pointer, as the memory pool which uses a user provided USM allocation, will also take a size, so the memory pool will assume this size is the correct size of the allocation. If this is not the case then the result is undefined behaviour.

The reason the `initial_threshold` property is set is that this property instructs the SYCL RT to maintain this amount of memory and not release it (if possible), and this behaviour is guaranteed by using an existing allocation.

Author: I've added wording which describes that modifying the ptr whilst the memory pool owns it is UB, see https://github.com/intel/llvm/pull/14800/commits/a7bcd952714d62ca2663ea24a6037bd4a8d1c994.

 `memory_pool` constructor to throw an exception with the `errc::invalid` error
 code.

|===

=== Default memory pools

As well as being able to construct a memory pool explicitly, this extension
introduces a default memory pool per device for each SYCL context and device
pair for device allocations and a default memory pool per context for host
allocations.

New member functions are added to the `context` class to retrieve the default
memory pool as a copy of the `memory_pool` object. This can be modified and have
those modifications reflected as it conforms to the SYCL common reference
semantics.

[source,c++]
----
class context {

  memory_pool context::ext_oneapi_get_default_memory_pool() const;

  memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;

}; // context
----

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool() const;
----

_Returns_: The default memory pool associated with the context for allocating
with the allocation kind `usm::alloc::host`.

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;
----

> _Returns_: The default memory pool associated with the context and `dev` for

Author: That's a good point, I've clarified in https://github.com/intel/llvm/pull/14800/commits/f53598e0d8f33105d00fad526454dc08ab56b762 that an error should be thrown if the specified device is not associated with the context.

allocating with the allocation kind `usm::alloc::device`.

=== Asynchronous malloc & free

This extension introduces a series of new enqueue functions for enqueueing
asynchronous malloc and free commands which operate with the memory pools also
introduced in this extension.

All enqueue functions introduced have overloads which take a SYCL `queue` and a
SYCL `handler`. None of enqueue functions return a SYCL `event` directly, as
this extension is in line with the
link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
calling `submit_with_event` and the `handler` overloads of these enqueue
functions.

[source,c++]
----
namespace ext::oneapi::experimental {

> void *async_malloc(queue q, size_t size);

Reviewer 5: It seems that this API is only useful for in order queues. Should we change this entry point to also take a vector of `event`s?

Author: This is following the convention of the new [enqueue functions extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc) where events are not provided or returned by default, to allow the implementation to be more efficient.This can still be used for both out-of-order queues as well as commands can be synchronised with the `barrier` or `partial_barrier` APIs and events can be used via the `submit_with_event` API.

void *async_malloc(handler h, size_t size);

> void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);

Reviewer 5: Same as above.

void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);

> void async_free(queue q, void *ptr);

Reviewer 5: Same as above

Author: See response above.

freed back to the source at this point.

When a SYCL queue which enqueued asynchronous malloc commands is synchronized
with, i.e. when `queue::wait` or `queue::wait_and_throw` is called, the SYCL RT
will check the current total memory allocated to the memory pool against the
memory pool's deallocation threshold. If the total memory allocated is larger
than than the threshold then the memory will be freed back to the source,
otherwise it will not.

[_Note:_ An implementation should maintain memory allocated to a memory pool to
the size specified by the deallocation threshold when possible, however, an
implementation is permitted to release this memory back to the source if it is
required by another memory pool or an application in another process.
_{endnote}_]

== Implementation notes

It is expected that for L0 this extension will be implemented within the L0
adapter, by reserving allocations for the memory pool and opportunistically
re-using the memory allocated based on the command lists being enqueued to the
L0 driver.

It is expected that for CUDA this extension will be implemented by mapping onto
the CUDA stream-ordered allocator feature.

== Issues

> . Should we allow mixing asynchronous and synchornous memory commands?

Reviewer 5: ```suggestion
. Should we allow mixing asynchronous and synchronous memory commands?
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/a7bcd952714d62ca2663ea24a6037bd4a8d1c994.

    void *temp = async_malloc_from_pool(q, 1024, memPool);

    // memory allocation is used for first kernel
    parallel_for(q, range{1024}, [=](id<1> idx) {
      do_something(idx, temp);
    });

    // memory is available to be used by another allocation
    async_free(q, temp);
  }

  {
    // memory pool re-uses previously allocated memory
    void *temp = async_malloc_from_pool(q, 1024, memPool);

    // memory allocation is used for second kernel
    parallel_for(q, range{1024}, [=](id<1> idx) {
      do_something_else(idx, temp);
    });

    // memory is available to be used by another allocation
    async_free(q, temp);
  }

  // memory pool releases memory back to the OS
  q.wait();
}
----

The above example does not use SYCL events, so below is the same example using
an out-of-order SYCL queue and SYCL events to manage dependencies.

[source,c++]
----
int main(int argc, char *argv[])
{
  queue q;

  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
    q.get_device(), usm::alloc::device);
  
  {
    void *temp = null;

    // memory pool allocates memory from the OS
    auto e1 = q.submit_with_event([&](handler &cgh) {

>       temp = async_malloc_from_pool(q, 1024, memPool);

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

    });

    // memory allocation is used for first kernel
    auto e2 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e1);

>       parallel_for(q, range{1024}, [=](id<1> idx) {

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

        do_something(idx, temp);
      });
    });

    // memory is available to be used by another allocation
    auto e3 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e2);

>       async_free(q, temp);

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

    });
  }

  {
    void *temp = null;

    // memory pool re-uses previously allocated memory
    auto e4 = q.submit_with_event([&](handler &cgh) {

>       temp = async_malloc_from_pool(q, 1024, memPool);

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

    });

    // memory allocation is used for second kernel
    auto e5 = q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e4);

>       parallel_for(q, range{1024}, [=](id<1> idx) {

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

        do_something_else(idx, temp);
      });
    });

    // memory is available to be used by another allocation
    q.submit_with_event([&](handler &cgh) {
      cgh.depends_on(e5);

>       async_free(q, temp);

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

    // memory allocation is used for second kernel
    parallel_for(q, range{1024}, [=](id<1> idx) {
      do_something_else(idx, temp);
    });

    // memory is available to be used by another allocation
    async_free(q, temp);
  }

  // again memory pool does not release memory back to the OS
  q.wait();
}
----

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
table below.  Applications can test for the existence of this macro to determine
if the implementation supports this feature, or applications can test the
macro's value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Memory pool

This extension introduces the `memory_pool` class, which provides a handle to a
memory pool owned by the SYCL runtime or a specific backend, and adheres to the

> SYCL common reference semantics.

Reviewer 7: Should it also support APIs to get / read the properties that are used to construct the pool?

Author: Yeah I think we should have get/set APIs for the properties, for SYCL 2020 properties, there are common property member functions which we add to any class which can be constructed with properties, but since we're using the properties extension I'm not actually sure what the convention is for doing this, is there a good example of this I could follow?

Reviewer 7: @aelovikov-intel, @Pennycook, and @rolandschulz are reviewing the new properties extension right now.  Can one of you propose the right convention for a class like `memory_pool` to follow?  This class takes several properties, but they are all runtime values.  What APIs should `memory_pool` expose in order to query them?  I assume it is some form of `has_property` and `get_property`?

Reviewer 8: I think we have a choice here.

We could either expose `has_property()` and `get_property()`:

```c++
auto pool = memory_pool(..., sycl::properties{sycl::max_size(1024)});
if constexpr (pool.has_property<sycl::max_size>()) {
  assert(pool.get_property<sycl::max_size>().value == 1024);
}
```

...or we could expose a `properties()` member that returns the property list that was passed to the constructor:

```c++
auto pool = memory_pool(..., sycl::properties{sycl::max_size(1024)});
auto p = pool.properties();
if constexpr (p.contains<sycl::max_size>()) {
  assert(p.get<sycl::max_size>().value == 1024);
}
```

...and yes, I did just suggest that we replace `get_property` and `has_property` with less verbose alternatives!

I'd prefer something like the second approach, because:

1) We don't have to specify `has_property()` and `get_property()` in every class.
2) Returning a `properties` allows the developer to do anything they can do with a `properties`, not just two queries.
3) We can adopt shorter names: `has_property()`/`get_property()` make sense when we're querying an arbitrary class, but if we know we're querying a property list, we can use something like `contains`/`get`.

(3. would be longer if you're only querying one property, but much shorter if you query multiple properties or want to call some function/trait that accepts a list of properties.)

Reviewer 9: I'm not so sure about `has`/`get` if we decide to go with duck-typing for a single property passed instead of the properties list. These names are much more likely to conflict with some specific property's methods.

Reviewer 9: > This class takes several properties, but they are all runtime values.

Why are they? It looks like `ready_only` and `zero_init` should be compile-time only properties (which we define differently). Are we going to provide variable shortcuts for these two or not?

Author: Was there a decision made on this as to a convention to follow? I like option 2, having a single function which returns the `properties` object.

If we were to go in this direction, I imagine we'd want to add a member function to `memory_pool` which looks something like:

```
properties<> properties() const;
```

Author: > Why are they? It looks like ready_only and zero_init should be compile-time only properties (which we define differently). Are we going to provide variable shortcuts for these two or not?

So these properties could be compile-time but I think they should be runtime properties as we only need know about these at runtime.

Reviewer 7: The use of properties in `memory_pool` is pretty normal.  If we can't use the new-style properties in this case, it points to a pretty serious flaw in the design.

Why can't we define `has_property` and `get_property`?  Wouldn't it be something like:

```
template<typename PropertyT>
bool has_property(PropertyT prop);

template<typename PropertyT>
PropertyT::type get_property(PropertyT prop);
```

I don't think this would cause any ABI break if we add new properties in the future, right?

Author: So we had some discussions about this offline, we could potentially use `has_property()` and `get_property()` as you describe, and this would work, but it was suggested that with the new properties mechanism there is benefit of having a single function which returns the `properties` object, and we discovered that this doesn't not work well with a class which isn't templated on the properties. A possible solution to this would be to have a polymorphic properties type, and we discussed that the SYCL 2020 `property_list` could potentially be used for this.

Reviewer 7: > A possible solution to this would be to have a polymorphic properties type, and we discussed that the SYCL 2020 `property_list` could potentially be used for this.

I don't see how we could use `property_list` exactly because `property_list` is expected to contain the old-style SYCL 2020 properties, not the new-style ones.  I guess we could create a new polymorphic type that contains the new-style properties.

It seems like your change 662ae8e38268c68789ca443b7ed22c60c0bc33b3 removes the new-style properties entirely and uses the old-style SYCL 2020 ones.  This seems a little drastic, but I guess it could be OK.  Is this just a temporary thing that we would remove later once we figure out the right pattern for querying the new-style properties?  Have you analyzed the change in 662ae8e38268c68789ca443b7ed22c60c0bc33b3 to see how disruptive it will be to users to switch back to compile-time properties?

It seems to me like it would be less disruptive to just add the `has_property` and `get_property` APIs now.  Then, we can remove just these two APIs later if we decide that it makes more sense to have a single `get_properties` function.

Author: So we didn't delve into this too far but the idea was that we could make the the new `properties` class convertible to the the SYCL 2020 `property_list` and have that behave as the polymorphic properties type. the benefit of this approach was that it could potentially reduce the impact of transitioning from SYCL 2020 properties to the new properties mechanism.

We seemed to leaning towards having a single API for retrieving properties (as opposed to individual `has_property` and `get_property`) which isn't currently workable and the idea of having `property_list` being a polymorphic type for the new properties mechanism seemed to have benefits. So when I was considering the options for this extension, I concluded that in either case we would need to use the `get_property` / `set_property` interface, which will need to change so there's no way to avoid an ABI/API break. Between the options sticking with the SYCL 2020 properties until we conclude on the solution for a polymorphic properties class in the new mechanism seemed to be less disruptive. Though this really depends on the conclusion of the new properties mechanism, so perhaps this will end up being more disruptive.

Reviewer 7: Recording the conclusion to our offline conversation here:

* Change the API back to use compile-time properties
* Do not provide any way to query the properties in this revision of the spec.  We can add that later once we settle the open issues discussed in the comments above.

Author: API updated to use the compile-time properties extension in https://github.com/intel/llvm/pull/14800/commits/0c87d8f9026daa9a3fbb6d6751787fd4ef2a0f90.

Reviewer 7: Marking this thread as resolved.  There is an issue in the spec noting that we still need to add an API to query the properties.

}; // memory_pool

}  // ext::oneapi::experimental
----

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(context ctx, Properties props = {});
----

_Effects_: Constructs a memory pool associated with `ctx` and all SYCL devices
associated with it, with the allocation kind `usm::host` and applying any
properties in `props`.

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
----

_Effects_: Constructs a memory pool associated with `ctx` and `dev`, with the
allocation kind `kind` and applying any properties in `props`.

_Throws_: An exception with the `errc::invalid` error code if `kind` is
`usm::host`.

[source, c++]
----
~memory_pool();
----

_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
pool to be destroyed after all remaining allocations have been freed, and
returns immediately without waiting.

[source, c++]
----
context get_context() const;
----

_Returns_: The SYCL context associated with the memory pool.

[source, c++]
----
std::vector<device> get_devices() const;
----

> _Returns_: The SYCL device(s) associated with the memory pool.

Author: So this depends on how the memory pol was created, if it is created for host allocations it will return all devices associated with the context, however, if it is created for device or shared allocations it will return the specific device that was specified during construction only.

> _Returns_: The SYCL device(s) associated with the memory pool.

Reviewer 7: This is not clear from the description, but I'm not sure this is the right API.  Do we really want to return a vector of devices here?  It seems like applications will really only care about the case when the pool contains "shared" or "device" memory, and there is just one device in those cases.

For comparison, consider the function `sycl::get_pointer_device`, which returns just a single device:

```
device get_pointer_device(const void* ptr, const context& syclContext)
```

I think we could do something similar for the memory pool.  I see two options:

* For "shared" and "device" USM pools, return the device that is specified in the pool constructor.  For "host" USM pools, arbitrarily return the first device in the context (same as `get_pointer_device`), or
* For "shared" and "device" USM pools, return the device that is specified in the pool constructor.  For "host" USM pools, throw an exception.

> _Returns_: The SYCL device(s) associated with the memory pool.

Author: I think I agree, my rationale for the current design was to return all devices, but this does mean creating a vector of devices so we should adopt one of the alternatives you suggested.

I would prefer to match the behaviour of `get_pointer_device` for consistency, I find it a little odd to return just the first device which is associated with the context, but in this case since the memory pool is associated with the whole context, the devices is not particularly relevant, so I think this is acceptable.

[source, c++]
----
usm::alloc get_alloc_kind() const;
----

_Returns_: The memory allocation kind of the memory pool.

[source, c++]
----
size_t get_max_size() const;
----

> _Returns_: The maximum size of the memory pool.

Reviewer 7: What happens if the `maximum_size` property wasn't provided?  For example, what value should we return for a USM "host" memory pool?  I'm not aware that we have any "max size" value to return in this case.

I wonder if it would be better to replace this query with a more general query that just tells the properties that were provided.

Author: So if the `maximum_size` property isn't provided then the memory pool will adopt an implementation defined max size. For USM host allocations where there isn't really a max size I would expect this to be something like `std::numeric_limits<size_t>::max()`.

Reviewer 7: I think there needs to be more clarity about what "max size" means.

I think the behavior would be clear if the `memory_pool` constructor actually allocated the "max size" memory.  This way, the application would have a guarantee that this amount of memory is available to the pool throughout its lifetime.  However, I don't think that's how you intent it do work.

If you don't preallocate the memory, how can you guarantee that the pool actually has "max size" memory available?  For example, what if there are two memory pools for the same device, where each specifies a "max size"?  Is there some check that the total of all the max sizes is less that the device's limit?  What if some of the device's memory is allocated outside of all pools?  This will decrease the amount of memory available to the pools for that device.  I don't see how you can guarantee that a pool will have "max size" memory available to it in these cases unless you preallocate the memnory.

Reviewer 6: This is a very good question, I think. For example, what happens if `async_malloc_from_pool` fails to allocate memory to the pool, but max size is not reached yet? Current description doesn't specify an error code for that

Author: So the maximum size is not guarantee that the pool will provide this amount of memory, instead it is a user provided upper limit to how much that memory pool can reserve. So even if there is not enough memory available on the device to allocate for an `async_malloc_from_pool` because the memory is allocated else ware the runtime can throw an error in this case, same as it would for a regular `malloc_device`.

Reviewer 7: This seems like a strange feature to have.  It seems like `maximum_size` just adds an artificial upper bound on the amount of memory that can be allocated from the pool.  Can you remind me who wants this feature and why?

Author: So the rationale for this feature is to provide control over memory allocation and encourage more efficient re-use of memory. This has been requested by oneAPI libraries in order to allow users or high level frameworks to control how much of the GPU memory is available to each library or set of operations.

Reviewer 7: Can you explain in more detail how you expect an application to use this?  Based on my current understanding, this is how I think it would work:

* Application constructs the memory pool with a certain `maximum_size`, which is intended to constrain the amount of memory that is available.
* The oneAPI libraries use this memory pool.  If they exceed the `maximum_size`, then they fail with an error because they can't allocation enough memory.

How is this useful to the application?  The application doesn't really control how much memory the oneAPI libraries use.  It just causes an error if they use more than the threshold.

Author: So your description of how the property would work is correct, however, the implementation doesn't necessarily have to fail as soon as `async_malloc_from_pool` requests an allocation which would exceed the maximum size. The memory pool has options such as re-allocating free memory already reserved by the pool or waiting on dependent free commands to complete to free up additional memory.

So while potentially failing to allocate memory due to the maximum size threshold being exceeded may not sound desirable, there are some benefits of this, provided it's used carefully:

* Firstly it allows isolation of resources, where the memory can be divided up between different GPU instances, processes, threads, libraries, etc, to avoid overallocation of memory in individual pools and memory being mapped back and forth between pools.
* Secondly it encourages the implementation to prioritise re-using free memory already reserved by the pool or waiting for dependent free commands, than continuing to grow the pool's reservation.
* Finally it can be useful as a development tool, if the application knows exactly how much memory is required, it can be used to detect memory leaks or unexpected increases to memory allocations.

This is also a feature of CUDA, to which we've aligned pretty closely with. Though CUDA does go slightly further with this and has additional attributes for specifying memory pool allocation strategies, for example in CUDA the default behaviour is to fail if all attempts to allocate memory are unsuccessful, but if you set the `cudaMemPoolReuseAllowOpportunistic` attribute on the pool, then it will instead allocate global memory directly.

Reviewer 7: > This is also a feature of CUDA, to which we've aligned pretty closely with.

Do CUDA memory pools guarantee that all of the memory in the pool is available once the pool is created?  This makes a lot of sense to me, and it's what I originally thought memory pools were for.  You create the pool up front.  If pool creating succeeds, then you know that all that memory will definitely be available later when you allocate from the pool.

However, the proposed SYCL memory pools do not work like this.  Allocating a pool of N bytes doesn't provide any guarantee that N bytes of memory are really available.

[source, c++]
----
size_t get_threshold() const;
----

_Returns_: The deallocation threshold of the memory pool.

[source, c++]
----
void set_new_threshold(size_t newThreshold);
----

_Effects_: Sets the deallocation threshold of the memory pool if the value of
`newThreshold` is larger than the current threshold.

_Throws_: An exception with the `errc::invalid` error code if the value of
`newThreshold` is lower than the current threshold or larger than the maximum
allocation size.

=== Memory pool properties

A memory pool can be constructed with a number of properties which can change
certain behaviors, these can be specified when constructing a `memory_pool`
object.

[source,c++]
----
namespace ext::oneapi::experimental {

struct initial_threshold_key {
  using value_t = property_value<initial_threshold_key>;

  initial_threshold_key(size_t initialThreshold);
};

struct maximum_size_key {
  using value_t = property_value<maximum_size_key>;

>   initial_threshold_key(size_t maxSize);

Reviewer 7: ```suggestion
  maximum_size_key(size_t maxSize);
```

Author: Fixed in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db.

};

struct read_only_key {
  using value_t = property_value<read_only_key>;

  read_only_key(bool readOnly);
};

struct zero_init_key {
  using value_t = property_value<zero_init_key>;

>   read_only_key(bool zeroInit);

Reviewer 7: ```suggestion
  zero_init_key(bool zeroInit);
```

Author: Fixed in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db.

};

struct use_existing_memory_key {
  using value_t = property_value<use_existing_memory_key>;

  use_existing_memory(void *ptr, size_t size);
};

inline constexpr initial_threshold_key::value_t initial_threshold;
inline constexpr maximum_size_key::value_t maximum_size;
inline constexpr read_only_key::value_t read_only;
inline constexpr zero_init_key::value_t zero_init;

> inline constexpr use_existing_memory_key::value_t use_existing_memory;

Reviewer 7: These `inline constexpr` variables are not correct for a runtime property.  See [Representation of Properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_properties.asciidoc#82-representation-of-properties) in the properties spec.

Author: Thanks, I've fixed those in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db.

}  // ext::oneapi::experimental
----

|===
|Property|Description

|`initial_threshold`
|The `initial_threshold` property specifies the initial deallocation threshold
 value for the memory pool. If this property is not used the default value is
 zero, and this can be increased after the memory pool is created by calling
 `memory_pool::set_new_threshold`.

|`maximum_size`
|The `maximum_size` property specifies the maximum size of the memory pool,
 after which any allocation will result in an exception. If the value specified
 is larger than the implementation can support an exception with the
 `errc::memory_allocation` error code is thrown. If this property is not used
 the default value is implementation-defined.

|`read_only`
|The `read_only` property is a performance hint which asserts that all memory
 allocations from the memory pool will only ever be read from, this can be used

>  by the SYCL runtime to optimize for performance.

Author: Good catch, this description is wrong, it should say that allocations will only be read from within kernels, so you can still write to the memory on the host.

|`zero_init`
|The `zero_init` property adds the requirement that all memory allocated to the
 memory as it is allocated from the source to the memory pool will be
 initialised to zero. Note there is no guarantee that the memory allocation be
 re-initialized to zero when it is re-allocated from the pool, so users must
 re-initialize memory to zero if they wish for later allocations to have this

>  behavior.

Reviewer 7: I wonder if we really need this property.  It seems like applications can define this semantic themselves by using `use_existing_memory` and zero-initializing the content of the buffer.

Author: That's true, you can achieve the zero-init semantics by using the `use_existing_memory` property and zero-initialising the data before passing the pointer to the memory pool, however, using the `use_existing_memory` also means the memory pool looses the ability to dynamically scale the amount of memory allocated as required, which is another important requirement.

Reviewer 7: OK.  I can see that.

|`use_existing_memory`
|The `use_existing_memory` property adds the requirement that the memory pool
 will use an existing USM memory allocation provided by the user instead of
 allocating from the OS. This property takes a pointer to a valid USM memory
 allocation of the same allocation kind as the memory pool is initialized with
 and the size of that memory allocation. Using this property will implicitly set
 the `maximum_size` and `initial_threshold` property to be that of the size
 provided, and as such using the `maximum_size` or `initial_threshold`
 properties in conjunction with this property will cause the `memory_pool`

>  constructor to throw an exception with the `errc::invalid` error code.

Reviewer 7: I like the `use_existing_memory` concept, but I wonder if it should be exposed as a new constructor instead.  There is no need for the pool to know the USM memory "kind" if the user provides their own USM memory, so the constructor could be simplified to:

```
template <typename Properties = empty_properties_t>
memory_pool(context ctx, void* ptr, size_t size, Properties props = {});
```

However, this makes it hard to implement `get_alloc_kind` in this case.  Is that query really useful, though?  I think I'd add the constructor I list above and also remove `get_alloc_kind`.

Author: I agree it would be better to expose this as an alternative constructor, I've switched from the property to a constructor like what you proposed in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db.

Though I don't think this necessarily makes it difficult to implement `get_alloc_kind`, I think the constructor which takes an existing USM memory allocation could infer the allocation kind by calling `get_pointer_type`.

Reviewer 7: > Though I don't think this necessarily makes it difficult to implement `get_alloc_kind`, I think the constructor which takes an existing USM memory allocation could infer the allocation kind by calling `get_pointer_type`

Good point.

|===

=== Default memory pools

As well as being able to construct a memory pool explicitly, this extension
introduces a default memory pool per device for each SYCL context and device
pair for device allocations and a default memory pool per context for host

> allocations.

Reviewer 7: There is also USM "system" memory.  Is there a default memory pool for that?

> allocations.

Author: There's currently not default memory pool for shared or system allocations, we could add this if there is a desire for it though.

> allocations.

Reviewer 7: What was the rationale for adding default pools for "device" and "host" but not "shared"?

> allocations.

Author: The rationale for this was simply that the motivational use cases were for host or device allocations and not shared. We considered adding default memory pools for shared allocations as well, we added an open issue for it at the end of the extension. After thinking about this some more, there's no reason not to include this as well, so I've altered the interface to allow for shared allocations as well in https://github.com/intel/llvm/pull/14800/commits/48493ace0babceba6b65fe7399b43284c10681db.

> allocations.

Author: On the question of system USM, we could potentially add this, it could be useful to provide a way to allocate system USM asynchronously, though it may not be so straight forward as system USM uses the system malloc/free so there's no `usm::alloc` enumeration for system USM.

> allocations.

Reviewer 7: Yes, I agree that it makes sense to omit "system" USM for now.

New member functions are added to the `context` class to retrieve the default
memory pool as a copy of the `memory_pool` object. This can be modified and have
those modifications reflected as it conforms to the SYCL common reference
semantics.

[source,c++]
----
class context {

  memory_pool context::ext_oneapi_get_default_memory_pool() const;

  memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;

}; // context
----

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool() const;
----

_Returns_: The default memory pool associated with the context for allocating
with the allocation kind `usm::host`.

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;
----

_Returns_: The default memory pool associated with the context and `dev` for
allocating with the allocation kind `usm::device`.

=== Asynchronous malloc & free

This extension introduces a series of new enqueue functions for enqueueing
asynchronous malloc and free commands which operate with the memory pools also
introduced in this extension.

All enqueue functions introduced have overloads which take a SYCL `queue` and a

> SYCL `handler`. None of enqueue functions return a SYCL `event` direction, as

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

this extension is in line with the
link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
calling `submit_with_event` and the `handler` overloads of these enqueue
functions.

[source,c++]
----
namespace ext::oneapi::experimental {

void *async_malloc(queue q, size_t size);

void *async_malloc(handler h, size_t size);

void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);

void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);

void async_free(queue q, void *ptr);

void async_free(handler h, void *ptr);

}  // ext::oneapi::experimental
----

[source, c++]
----
void *async_malloc(queue q, size_t size);

void *async_malloc(handler h, size_t size);

void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);

void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
----

_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
will asynchronously allocate memory of size `size` in bytes, allocating from the
memory pool `pool` if provided, otherwise allocation from the default memory

> pool associated with the SYCL context and device associated with `q` or `h`.

Reviewer 7: When allocating from the default memory pool, how do you choose the type of USM memory (shared, host, or device)?

Author: Good catch, I think we are missing a `usm::alloc` parameter to `async_malloc` to specify which default pool to allocate from. Though if we're adding this to those functions and not the `async_malloc_from_pool` functions, perhaps we should maybe reconsider whether the allocation kind is specified on pool creation or during allocation.

Reviewer 7: The `async_malloc_from_pool` function needs to take a pool handle because there could be multiple pools that allocate the same "kind" of USM memory.

Author: Yeah in second thought I think it's fine to just have the allocation kind provided to `async_malloc`. I've updated the interface to reflect this in https://github.com/intel/llvm/pull/14800/commits/48493ace0babceba6b65fe7399b43284c10681db.

Memory is first allocated from the memory pool if possible, otherwise memory is
allocated from the source to the memory pool to provide enough memory in the

> memory pool for the allocation.

Author: I think this needs to be wording in the spec as this behaviour is what guarantees the benefit of using the memory pool, the "if possible" part does provide flexibility to an implementation to decide when this happens and when it doesn't.

> memory pool for the allocation.

Reviewer 7: I'm still not convinced this should be in the spec, especially since we say "if possible".  If it's not a hard requirement, how can you test for it?  This sounds more like something that is left to the implementation as a "quality of implement" choice.

> memory pool for the allocation.

Author: I've re-written the rules for the memory pool to be less prescriptive in https://github.com/intel/llvm/pull/14800/commits/044ec4427945bcbd526ed3874e0989fe8ffa54e7, hopefully this addresses this.

_Returns_: A pointer to the address of a memory reservation.

> _Throws_: An exception with the `errc::invalid` error code if `size` is zero.

Author: That makes sense, I'll change this to match those.

An exception with the `errc::memory_allocation` error code if the allocation
brings the memory pool over the maximum size. This error can be thrown

> asynchronously.

Author: Yeah I agree it makes sense to just require that this is an async error, changed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.

[_Note:_ Accessing the memory at the address of the pointer returned by
asynchronous malloc functions before the command has completed execution is

> undefined behavior. _{endnote}_]

Author: Good catch, fixed in https://github.com/intel/llvm/pull/14800/commits/48493ace0babceba6b65fe7399b43284c10681db.

[source, c++]
----
void async_free(queue q, void *ptr);

void async_free(handler h, void *ptr);
----

_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
will asynchronously free the memory allocation at the address of `ptr`. Memory
will be freed from the memory pool to be used by other asynchronous malloc
commands which execute later, and will not free to the source until the SYCL
queue associated with the asynchronous allocation command has been synchronized

> with.

Author: > Memory is returned back to the pool when the command containing the call to async_free completes.

So I don't think this is the wording that we want here, for the memory pool to be efficient we don't want the memory being released by the OS until you synchronise, since you want subsequent `async_malloc_from_pool` commands to be able to utilise the memory in the pool and not allocate from the OS.

Though I think I agree that this wording is overly prescriptive, instead of saying that memory will not be freed until you call `wait`, I think perhaps it's better to say that memory that is freed with `async_free` may be reused by a subsequent `async_malloc_from_pool` avoiding allocating from the OS, but the memory will be released by the OS when you call `wait`.

> Is the second async_malloc_from_pool guaranteed to succeed even though there is no intervening "wait" since the call to async_free?

So in this example, assuming `q` is in-order, the memory in the pool will be freed up to be used in the second kernel, it's not necessary to call `wait` that is only for the memory to be released back to the OS.

> with.

Reviewer 7: Why is it necessary for the spec to specify a difference between "freeing memory to the pool" vs. "releasing memory to the OS"?  Shouldn't this be an implementation detail?

> So I don't think this is the wording that we want here, for the memory pool to be efficient we don't want the memory being released by the OS until you synchronise

The spec should not mandate an algorithm just because it is more efficient.  Efficiency of design should be an implementation detail.  If you have thoughts about how an efficient design should work, it would be appropriate to add an "Implementation notes" section at the bottom of the spec, but that wording should be non-normative.

> with.

Author: I see what you mean, in the original wording I was aiming to describe the most optimal behaviour, but I agree, perhaps we should be as prescriptive, and leave more to implementation detail. I've re-written the memory pool rules to be less prescriptive in https://github.com/intel/llvm/pull/14800/commits/044ec4427945bcbd526ed3874e0989fe8ffa54e7.

_Throws_: An exception with the `errc::invalid` error code if `ptr` is not the

> address of a memory allocation allocated to a memory pool.

Author: That's true I've updated this to be undefined behaviour instead in https://github.com/intel/llvm/pull/14800/commits/48493ace0babceba6b65fe7399b43284c10681db.

implementation is permitted to release this memory back to the source if it is
required by another memory pool or an application in another process.
_{endnote}_]

== Implementation notes

It is expected that for L0 this extension will be implemented within the L0
adapter, by reserving allocations for the memory pool and opportunistically
re-using the memory allocated based on the command lists being enqueued to the
L0 driver.

It is expected that for CUDA this extension will be implemented by mapping onto
the CUDA stream-ordered allocator feature.

== Issues

. Should we allow mixing asynchronous and synchornous memory commands?
+
--
*UNRESOLVED*: CUDA allows memory allocated with the asynchronous malloc command
to be freed with the regular synchronous free command, should we extend this
capability to SYCL?
--

. Should we allow freeing memory with a different queue?
+
--
*UNRESOLVED*: Should we allow a memory allocation allocated with an asynchronous
malloc command from one queue to be freed by an asynchronous free command from
another queue?
--

. How should we treat an asynchronous malloc command with a memory pool that is

> not associated with that queue?

Reviewer 7: I don't understand this issue.  A memory pool is associated with a context, not with a queue.  Did you mean to ask about the case when the memory pool's context does not match the context of the queue?

Author: I've added clarification that `async_malloc` and `async_malloc_from_pool` require the queue to be associated with the same context and devices as the memory pool in https://github.com/intel/llvm/commit/48493ace0babceba6b65fe7399b43284c10681db, so I've removed this issue.

+
--
*UNRESOLVED*: If an asynchronous malloc command is enqueued with a memory pool
that is not associated with the queue that the command is enqueued from, should
this result in an error?
--

. Should we have a default memory pool for `usm::shared`?
+
--
*UNRESOLVED*: Currently the proposed API means that there cannot be default
memory pool for allocations of allocation kind `usm::shared`, and therefore a
user must create their own explicit memory pool to do so. Is this reasonable or
should we extend the API to include a default memory pool for allocations of

> allocation kind `usm::shared`?

Reviewer 7: I saw this issue after making my comment above about the memory pool for "shared".  Why wouldn't we do this?  It seems weird to provide default memory pools for "host" and "device" but not "shared".

= sycl_ext_oneapi_async_memory_alloc

:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en
:dpcpp: pass:[DPC++]
:endnote: &#8212;{nbsp}end{nbsp}note

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2024 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

> This extension is written against the SYCL 2020 revision 8 specification.  All

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/f53598e0d8f33105d00fad526454dc08ab56b762.

references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

This extension also depends on the following other SYCL extensions:

* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
  sycl_ext_oneapi_properties]
* link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions]
* link:../supported/sycl_ext_oneapi_default_context.asciidoc[
  sycl_ext_oneapi_default_context]

== Status

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Backend support status

> This extension is not yet implemented in {dpcpp}.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/f53598e0d8f33105d00fad526454dc08ab56b762.

queues.

[source,c++]
----
namespace ext::oneapi::experimental {

class memory_pool {

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(queue q, usm::alloc kind, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, void *ptr, size_t size, Properties props = {});

  ~memory_pool();

  context get_context() const;

  device get_device() const;

  usm::alloc get_alloc_kind() const;

  size_t get_max_size() const;

  size_t get_threshold() const;

  void set_new_threshold(size_t newThreshold);

}; // memory_pool

}  // ext::oneapi::experimental
----

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(context ctx, Properties props = {});
----

_Effects_: Constructs a memory pool associated with `ctx` and all SYCL devices
associated with it, with the allocation kind `usm::alloc::host` and applying any

> properties in `props`.

Author: I agree, if there is a `usm::alloc` parameter for the host constructor, while the only valid option would be `usm::alloc::host` it will make it clearer to readers of the code what kind of memory the pool will allocate, I've changed this in https://github.com/intel/llvm/pull/14800/commits/0ebc79b53a653b0e25b136a2d11cbd18db9da97a.

from `q` and `dev`, with the allocation kind `kind` and applying any properties
in `props`. If `kind` is `usm::alloc::host` this is equivelant to calling
`memory_pool(queue.get_context(), props)` otherwise this is equivelant to
calling `memory_pool(queue.get_context(), queue.get_device(), props)`.

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(context ctx, void *ptr, size_t size, Properties props = {});
----

_Effects_: Constructs a memory pool associated with `ctx` which uses an existing
USM memory allocation `ptr` of size (in bytes) `size`, and applying any
properties in `props`. The memory pool will use the existing USM memory
allocation instead of allocating from the SYCL runtime. The associated SYCL
device and allocation kind of the memory pool is inferred from the `ptr`. If
`ptr` is not a valid USM pointer of kind `usm::alloc::host`,
`usm::alloc::device` or `usm::alloc::shared` the result is undefined behavior.
The maximum size and initial threshold of the memory pool are set to `size`.

_Throws_: An exception with the `errc::invalid` error code if `props` contains

> the `maximum_size` or `initial_threshold` properties.

Reviewer 8: I'm not sure I understand why `max_size` and `initial_threshold` are properties at all.

It seems like there are dedicated queries for them (`get_max_size()` and `get_threshold()`).  Without any properties, the max size is set to an implementation-defined limit, and the threshold is set to zero.

Why not just use a regular overload here with some runtime values?
```c++
template <typename Properties = empty_properties_t>
memory_pool(queue q, usm::alloc kind, int initial_threshold = 0, int maximum_size = /* implementation-defined */, Properties props = {});
```

That seems like it would give the best error messages, and it wouldn't pull in the complexities of using properties.

> the `maximum_size` or `initial_threshold` properties.

Author: I agree, these properties could be provided as regular parameters to the constructor, however, these are intended to be optional and there are use cases where you may want to provide one or the other, and having them as regular parameters may require the user to specify one in order to specify the next. So I thought this would be a good place to use properties so users only need specify the specific parameters they need to. I also thinks this makes it more future proof to adding additional optional parameters in the future.

> the `maximum_size` or `initial_threshold` properties.

Reviewer 7: Getting back to my comment about the exception ... I think this error condition can be diagnosed at compile time, so there is no need to throw an exception.  You should be able to either add a SFINAE constraint or a static-assert.  See my comment above.

> the `maximum_size` or `initial_threshold` properties.

Author: Yeah I see what you mean, if we use the new properties mechanism then the use of the incompatible properties would be known at compile-time and since this limitation only applies to this overload we could detect this at compile-time, so this would be a constraint, rather than an exception.

> the `maximum_size` or `initial_threshold` properties.

Author: Now that the APi is now using the compile-time properties extension I've updated the restriction on allocation kinds to be a constraint rather than a runtime error, see https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

 functions, this can be used by the SYCL runtime to optimize for performance.

|`zero_init`
|The `zero_init` property adds the requirement that all memory allocated to the
 memory as it is allocated from the source to the memory pool will be
 initialised to zero. Note there is no guarantee that the memory allocation be
 re-initialized to zero when it is re-allocated from the pool, so users must
 re-initialize memory to zero if they wish for later allocations to have this
 behavior.

|===

=== Default memory pools

As well as being able to construct a memory pool explicitly, this extension
introduces a default memory pool per device for each SYCL context and device
pair for device allocations and a default memory pool per context for host
allocations.

New member functions are added to the `context` class to retrieve the default
memory pool as a copy of the `memory_pool` object. This can be modified and have
those modifications reflected as it conforms to the SYCL common reference
semantics.

[source,c++]
----
class context {

  memory_pool context::ext_oneapi_get_default_memory_pool() const;

  memory_pool context::ext_oneapi_get_default_memory_pool(device dev,

>     usm::alloc kind) const;

Reviewer 7: I think this is not fixed.

}; // context
----

[source, c++]
----

> memory_pool context::ext_oneapi_get_default_memory_pool() const;

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/602e9952aa27058866491d7d18e52b8d832ec1c7.

> memory_pool context::ext_oneapi_get_default_memory_pool() const;

Reviewer 7: Also does not seem to be fixed.

> memory_pool context::ext_oneapi_get_default_memory_pool() const;

Author: Sorry I misunderstood this comment before, it's actually fixed now in https://github.com/intel/llvm/pull/14800/commits/0ebc79b53a653b0e25b136a2d11cbd18db9da97a.

void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
----

_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
will asynchronously allocate memory of size `size` in bytes, allocating from the
memory pool `pool` if provided, otherwise allocation from the default memory
pool associated with the SYCL context and device associated with `q` or `h`.
If a memory pool `pool` is provided the allocation kind is that of the memory
pool, otherwise if an allocation kind `kind` parameter is provided, this is the
allocation kind which is used. If a memory pool `pool` is provided, this must be
associated with the same SYCL context and device as `q` or the SYCL queue
associated with `h`. Memory is first allocated from the memory pool if possible,
otherwise memory is allocated from the source to the memory pool to provide
enough memory in the memory pool for the allocation. Accessing the memory at the
address of the pointer returned by asynchronous malloc functions before the
command has completed execution is undefined behavior.

_Returns_: A pointer to the address of a memory reservation if `size` is

> non-zero, otherwise returns `nullptr`.

Reviewer 7: This does not match match the core SYCL behavior of the USM allocation functions.  The core spec says:

> When the allocation size is zero bytes (numBytes or count is zero), these functions behave in a manner consistent with C++ std::malloc. The value returned is unspecified in this case, and the returned pointer may not be used to access storage. If this pointer is not null, it must be passed to sycl::free to avoid a memory leak.

Reviewer 7: Just pointing out that this issue is still open.

Author: This is fixed now in https://github.com/intel/llvm/pull/14800/commits/0ebc79b53a653b0e25b136a2d11cbd18db9da97a.

  // still within the specified reserve threshold
  q.wait();

  {
    // memory pool re-uses previously allocated memory
    void *temp = syclex::async_malloc_from_pool(q, 1024, memPool);

    // memory allocation is used for second kernel
    syclex::parallel_for(q, range{1024}, [=](id<1> idx) {
      do_something_else(idx, temp);
    });

    // memory is available to be used by another allocation
    syclex::async_free(q, temp);
  }

  // again memory pool does not release memory back to the memory provider
  q.wait();
}
----

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
table below.  Applications can test for the existence of this macro to determine
if the implementation supports this feature, or applications can test the
macro's value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the

>  feature-test macro always has this value.

Reviewer 10: Should this extension be versioned if it is growing over time?  Or are experimental specs not versioned as they're expected to change?

* When the maximum size of the memory pool is set, the value specified may
  round up to the nearest multiple of an implementation defined allocation chunk
  size.

Memory pools will work with both in-order and out-of-order SYCL queues.

[source,c++]
----
namespace ext::oneapi::experimental {

class memory_pool {

  template <typename Properties = empty_properties_t>
  memory_pool(const context &ctx, const device &dev, usm::alloc kind,
    Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(const queue &que, usm::alloc kind, Properties props = {});

  ~memory_pool();

  context get_context() const;

  device get_device() const;

  usm::alloc get_alloc_kind() const;

  size_t get_threshold() const;

  size_t get_reserved_size_current() const;

  size_t get_used_size_current() const;

  void increase_threshold_to(size_t newThreshold);

}; // memory_pool

}  // ext::oneapi::experimental
----

[source, c++]
----
template <typename Properties = empty_properties_t>
memory_pool(const context &ctx, const device &dev, usm::alloc kind,
  Properties props = {});
----

_Effects_: Constructs a memory pool which is associated with the context `ctx`

> and device `dev`, which allocates either device or shared USM memory depending

Reviewer 7: This sentence also mentions "shared" USM memory.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/781c697155825026fcb412aba621277404c76901.

  dictates the type of USM memory that is allocated to the pool.
* A memory pool is associated with a context and one or more device(s),
  depending on the allocation kind.

A memory pool and the asynchronous malloc and free commands which interact with
them adhere to the following behaviors:

* When an asynchronous malloc command is enqueued a valid USM pointer of the
  allocation kind is returned immediately. The asynchronous malloc command may
  execute any time between the command being enqueued and the command being
  synchronized with. The pointer returned is not valid until the asynchronous
  malloc command has completed.
* When an asynchronous malloc command executes it will allocate USM memory of
  the allocation kind from the memory pool.
* When a USM pointer is being allocated from the memory pool, it may allocate
  from free memory in the pool, including memory which was previously allocated
  to a USM pointer and has been freed or it may opportunistically re-use memory
  allocated to an existing USM pointer, for which an asynchronous free command
  has been enqueued to free, and that the asynchronous malloc command is

>   dependent on.

Reviewer 7: This bullet seems like an implementation detail, not something that belongs in the spec.  It just lists some implementation options, and doesn't mandate anything.  How would you test for this?

Author: I agree, I've moved this to the implementation notes sections in https://github.com/intel/llvm/pull/14800/commits/2898549e9ad8c0f9c947536cf7f993cb44aa928b.

* Additional memory can be allocated to the memory pool when required for an
  asynchronous malloc command or otherwise. Allocations to the pool must be in a

>   multiple of the allocation chunk size.

Reviewer 7: Isn't this also talking about an internal implementation detail?

Author: I've moved this to the implementation notes sections in https://github.com/intel/llvm/pull/14800/commits/2898549e9ad8c0f9c947536cf7f993cb44aa928b.

* When an asynchronous free command is enqueued, the USM pointer to be freed is
  immediately no longer valid. The asynchronous free command may execute any
  time between the command being enqueued and the command being synchronized
  with. Once the asynchronous free command is enqueued memory allocated to that
  USM pointer is now free for the memory pool to re-allocate to another USM
  pointer either once the command has completed, or if an asynchronous command
  is dependent on this asynchronous free command.
* The memory pool will not release memory in the pool that is currently
  allocated to a USM pointer, however, it may release any free memory in the
  pool. When the memory pool releases memory from the pool, this is done so in

>   multiples of the allocation chunk size.

Reviewer 7: This seems like an implementation detail too.

Author: I agree, I've moved this to the implementation notes sections in https://github.com/intel/llvm/pull/14800/commits/2898549e9ad8c0f9c947536cf7f993cb44aa928b.

* The memory pool should try to maintain an amount of memory in the pool
  set by the threshold size in order to better facilitate re-use of the memory
  between subsequent asynchronous malloc commends, however, it is not required
  to.
* The amount of total and free memory reported by the memory pool is updated
  immediately after returning from calls to enqueue asynchronous malloc and free

>   commands, even if the command does not execute immediately.

Reviewer 7: This makes me wonder how useful these queries are.  Is anyone asking for them?  It does not seem useful to provide a query about the amount of free memory if there is no guarantee that the value returned actually represent the amount of free memory at the time the call is made.  How would an application use such a query?

Author: This is a good point, this was a mistake, the amount of free memory should not be updated immediately on freeing instead it should be when the free command completes, otherwise as you said it could be reporting incorrectly how much memory is available in the pool, I've fixed this in https://github.com/intel/llvm/commit/0ebc79b53a653b0e25b136a2d11cbd18db9da97a.

  round up to the nearest multiple of the allocation chunk size.

Memory pools are intended to be used for both in-order and out-of-order SYCL
queues.

[source,c++]
----
namespace ext::oneapi::experimental {

class memory_pool {

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(queue q, usm::alloc kind, Properties props = {});

  template <typename Properties = empty_properties_t>
  memory_pool(context ctx, void *ptr, size_t size, Properties props = {});

  ~memory_pool();

  context get_context() const;

  device get_device() const;

  usm::alloc get_alloc_kind() const;

  size_t get_allocation_chunk_size() const;

  size_t get_max_size() const;

  size_t get_threshold() const;

  size_t get_current_size() const;

  size_t get_free_size() const;

  void set_new_threshold(size_t newThreshold);

}; // memory_pool

}  // ext::oneapi::experimental
----

[source, c++]
----
template <typename Properties = empty_properties_t>

> memory_pool(context ctx, Properties props = {});

Reviewer 7: I made this comment also in an earlier draft.  I think this constructor will be confusing because it is not obvious that it allocates "host" USM.  I think it might be better to add a `kind` parameter, and then also throw an exception if `kind` is not `usm::alloc::host`.

This also allows you to use this constructor in the future for `usm::alloc::shared` when there is no associated device.  Level Zero has the ability today to create "shared" USM memory that is not associated with a device, so it's possible we may add that ability to SYCL someday.

the `maximum_size` or `initial_threshold` properties.

[source, c++]
----
~memory_pool();
----

_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
pool to be destroyed after all remaining allocations have been freed, and
returns immediately without waiting.

[source, c++]
----
context get_context() const;
----

_Returns_: The SYCL context associated with the memory pool.

[source, c++]
----
device get_device() const;
----

_Returns_: The SYCL device associated with the memory pool. If the memory pool
allocates memory of kind `usm::alloc::device` or `usm::alloc::host::shared` the
device returned is the same device which was passed when constructing the memory
pool. If the memory pool allocates memory of kind `usm::alloc::host` the device
returned is the first of the devices associated with the context, which is
associated with the memory pool.

[source, c++]
----
usm::alloc get_alloc_kind() const;
----

_Returns_: The memory allocation kind of the memory pool.

[source, c++]
----
size_t get_allocation_chunk_size() const;
----

_Returns_: The allocation chunk size of the device associated with the memory

> pool.

Reviewer 7: Is it important to provide an API to query the chunk size?  It seems like this is an implementation detail.  The only affect on the API is with `set_new_threshold` and `maximum_size`.  It seems like `set_new_threshold` does not need to say precisely how the threshold value is rounded because the threshold is just a hint anyways.  For `maximum_size`, it seems like the wording could say something like:

> Because the memory pool may allocate memory in discrete chunks, the value may be rounded up to account for the size of these chunks.

It does not seem necessary for the application to know the chunk size in order to use this property.

Author: The motivation for being able to query the chunk size is due to the fact that implementations will round down the threshold or round up the maximum size, so if the user can check for this size, then they can choose more optimal sizes for these parameters.

[source, c++]
----
size_t get_max_size() const;
----

_Returns_: The maximum size of the memory pool.

[source, c++]
----
size_t get_threshold() const;
----

_Returns_: The deallocation threshold of the memory pool.

[source, c++]
----
size_t get_current_size() const;
----

_Returns_: The total memory allocated to the pool at the point the function is
called.

[source, c++]
----
size_t get_free_size() const;
----

_Returns_: The total free memory in the pool at the point the function is
called.

[source, c++]
----
void set_new_threshold(size_t newThreshold);
----

_Effects_: Sets the deallocation threshold of the memory pool if the value of
`newThreshold` is larger than the current threshold. If `newThreshold` is not a
multiple of the allocation chunk size, the new threshold is rounded down to the
nearest multiple.

_Throws_: An exception with the `errc::invalid` error code if the value of
`newThreshold` is lower than the current threshold or larger than the maximum

> allocation size.

Reviewer 7: Do you really want to throw an exception if the application attempts to lower the threshold?  What is the use case for this function?

Would it be better to name the function `increase_threshold_to` and then silently ignore the call if the threshold is already at (or greater than) `newthreshold`?

associated with `h`. Memory can either be allocated directly from the memory
pool or allocated to the memory pool to provide enough memory in the memory pool
for the allocation. Accessing the memory at the address of the pointer returned
by asynchronous malloc functions before the command has completed execution is
undefined behavior.

_Returns_: A pointer to the address of a memory reservation if `size` is
non-zero, otherwise returns `nullptr`.

_Throws_: An exception with the `errc::memory_allocation` error code if the
allocation brings the memory pool over the maximum size. This error must be
thrown asynchronously. An exception with the `errc::invalid` error code if
`kind` is `usm::alloc::unknown`.

[source, c++]
----
void async_free(queue q, void *ptr);

void async_free(handler h, void *ptr);
----

_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
will asynchronously free the memory allocation at the address of `ptr`. Memory
will be freed from the memory pool to be used by other asynchronous malloc
commands which execute later, and will not free until the SYCL queue associated

> with the asynchronous allocation command has been synchronized with. Accessing

Reviewer 7: This part of the sentence seems overly proscriptive:

> and will not free until the SYCL queue associated with the asynchronous allocation command has been synchronized with

I think you are trying to draw some distinction between "when the memory is freed" and "when the memory is released to the O/S".  As the first part of the sentence says, the memory is freed back to the pool as soon as the command completes.  I do not think the spec should say when the memory is released back to the O/S.

  q.wait();
}
----

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
table below.  Applications can test for the existence of this macro to determine
if the implementation supports this feature, or applications can test the
macro's value to determine which of the extension's features the implementation
supports.

=== Querying async memory alloc support

The following device query is provided to query whether a SYCL implementation
provides support for async memory allocation on a particular device.

The device aspects for this query is:

[frame="none",options="header"]
|======================
|Device descriptor |Description
|`aspect::ext_oneapi_async_memory_alloc` | Indicates if the device supports 
asynchronous memory allocation and memory pools.
|======================

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.

> |===

Author: Oh yeah, I think this ended up in the next section when I added the aspect, fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

  which is reserved by the pool from the memory provider.
* A memory pool has a *used size* (in bytes) which tracks the total memory which
  allocated has been allocated to a USM pointer.
* A memory pool has a *allocation kind* which is a value of `usm::alloc` which
  dictates the type of USM memory that is allocated to the pool.
* A memory pool is associated with a context and one or more device(s),
  depending on the allocation kind.

A memory pool and the asynchronous malloc and free commands which interact with
them adhere to the following behaviors:

* When an asynchronous malloc command is enqueued a valid USM pointer of the
  allocation kind is returned immediately. The asynchronous malloc command may
  execute any time between the command being enqueued and the command being
  synchronized with. The pointer returned is not valid until the asynchronous
  malloc command has completed.
* When an asynchronous malloc command executes it will allocate USM memory of
  the allocation kind from the memory pool.
* When an asynchronous free command is enqueued, the USM pointer to be freed is
  immediately no longer valid. The asynchronous free command may execute any

>   time between the command being enqueued and the command being synchronized

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

>   with. Once the asynchronous free command is enqueued memory allocated to that

Author: So this is intentionally specified as when enqueued, as it leave the option for an implementation to re-use previously allocated memory, providing this allocation is dependent on the freeing of that memory.

>   with. Once the asynchronous free command is enqueued memory allocated to that

Reviewer 7: If you leave it as "is enqueued", then this statement seems to be contradictory to the one before that says "the USM pointer to be freed is no longer valid once the dependencies of that command are satisfied".

A command "is enqueued" when the application calls `async_free` from `submit`, and this happens even before dependencies are satisfied.

I think this last sentence trying to describe an optimization where the implementation can reuse a freed pointer in a subsequent command that calls `async_malloc`.  I agree that's a valid optimization, but I think the wording here is not quite right.  You don't want to give the impression that the  optimization can reuse the memory even before the command calling `async_free` has its dependencies satisfied.

Perhaps the best solution is just to remove this sentence.  I think it is not necessary to describe this optimization in the spec because it seems like an implementation detail.

[source,c++]
----
namespace ext::oneapi::experimental {

class memory_pool {

  memory_pool(const context &ctx, usm::alloc kind,
    const property_list& propList = {});

  memory_pool(const context &ctx, const device &dev, usm::alloc kind,
    const property_list& propList = {});

  memory_pool(const queue &que, usm::alloc kind,
    const property_list& propList = {});

  memory_pool(const context &ctx, void *ptr, size_t size,
    const property_list& propList = {});

  ~memory_pool();

  context get_context() const;

  device get_device() const;

  usm::alloc get_alloc_kind() const;

  size_t get_threshold() const;

  size_t get_reserved_size_current() const;

  size_t get_used_size_current() const;

  void increase_threshold_to(size_t newThreshold);

  /* -- SYCL 2020 property interface members -- */

}; // memory_pool

}  // ext::oneapi::experimental
----

[source, c++]
----
memory_pool(const context &ctx, usm::alloc kind,
  const property_list& propList = {});
----

_Effects_: Constructs a memory pool which is associated with the context `ctx`
and all SYCL devices associated with it, with the allocation kind `kind` and

> applying any properties in `propList`.

Author: I agree, fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

_Throws_: An exception with the `errc::invalid` error code if `kind` is
`usm::alloc::device`, `usm::alloc::shared` or `usm::alloc::unknown`.

[source, c++]
----
memory_pool(const context &ctx, const device &dev, usm::alloc kind,
  const property_list& propList = {});
----

_Effects_: Constructs a memory pool which is associated with the context `ctx`
and device `dev`, with the allocation kind `kind` and applying any properties in

> `propList`.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

`usm::alloc::unknown`.

[source, c++]
----
memory_pool(const context &ctx, void *ptr, size_t size,
  const property_list& propList = {});
----

_Effects_: Constructs a memory pool which is associated with context `ctx`, 
uses an existing USM memory allocation `ptr` of size (in bytes) `size`,
and applying any properties in `propList`. The memory pool will use the existing
USM memory allocation instead of reserving memory from the memory provider. The
associated device and allocation kind of the memory pool is inferred from the
`ptr`. If `ptr` is not a valid USM pointer of kind `usm::alloc::host`,
`usm::alloc::device` or `usm::alloc::shared` or the value `size` does not match
the size of memory allocated for `ptr` the result is undefined behavior. If
`ptr` is freed or the memory it points to is written to whilst the memory pool
is still alive the result is undefined behavior. The maximum size and initial
threshold of the memory pool are set to `size`.

_Throws_: An exception with the `errc::invalid` error code if `propList`
contains the `maximum_size` or `initial_threshold` properties.

[source, c++]
----
~memory_pool();
----

> _Effects_: If this was the last copy, signals tto the runtime for the memory

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/d154a67861e46477ba07b415b845348f880ce8a4.

 zero, and this can be increased after the memory pool is created by calling
 `memory_pool::increase_threshold_to`.

|`maximum_size`
|The `maximum_size` property specifies the maximum size of the memory pool,
 after which any allocation will result in an exception. If the value specified
 is not a multiple of the allocation chunk size, the value is rounded up to the
 nearest multiple. If the value specified is larger than the implementation can
 support an exception with the `errc::memory_allocation` error code is thrown.
 If this property is not used the default value is implementation-defined.

|`read_only`
|The `read_only` property is an assertion by the application that allocations
from this pool will never be written from a kernel. Implementations may be able
to perform certain optimizations when a memory pool is created with this
property.

[Note: A memory pool created with the `read_only` property may still be written
from the host and may still be the destination of a memory copy operation that
is submitted to a device. -- end note]

|`zero_init`
|The `zero_init` property adds the requirement that all memory allocated to the

>  memory as it is allocated to the memory pool will be initialised to zero. Note

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/d154a67861e46477ba07b415b845348f880ce8a4.

 the memory allocation be re-initialized to zero when it is re-allocated from

>  the pool, so users must re-initialize memory to zero if they wish for later

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

 allocations to have this behavior.

|===

=== Default memory pools

As well as being able to construct a memory pool explicitly, this extension
introduces a default memory pool per device for each SYCL context and device
pair for device allocations and a default memory pool per context for host
allocations.

New member functions are added to the `context` class to retrieve the default
memory pool as a copy of the `memory_pool` object. This can be modified and have
those modifications reflected as it conforms to the SYCL common reference
semantics.

[source,c++]
----
class context {

  memory_pool context::ext_oneapi_get_default_memory_pool(usm::alloc kind)
    const;

  memory_pool context::ext_oneapi_get_default_memory_pool(const device &dev,
    usm::alloc kind) const;

}; // context
----

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool(usm::alloc kind) const;
----

_Returns_: The default memory pool associated with the context for allocating

> with the allocation kind `kind`.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

_Throws_: An exception with the `errc::invalid` error code if `kind` is
`usm::alloc::device`, `usm::alloc::shared` or `usm::alloc::unknown`.

[source, c++]
----
memory_pool context::ext_oneapi_get_default_memory_pool(const device &dev,
  usm::alloc kind) const;
----

_Returns_: The default memory pool associated with the context and `dev` for

> allocating with the allocations of kind `kind`.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

_Throws_: An exception with the `errc::invalid` error code if `dev` is not one
of the devices associated with the context. An exception with the
`errc::invalid` error code if `kind` is `usm::alloc::host` or
`usm::alloc::unknown`.

=== Asynchronous malloc & free

This extension introduces a series of new enqueue functions for enqueueing
asynchronous malloc and free commands which operate with the memory pools also
introduced in this extension.

All enqueue functions introduced have overloads which take a SYCL `queue` and a

> SYCL `handler`. None of enqueue functions return a SYCL `event` directly, as

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/d154a67861e46477ba07b415b845348f880ce8a4.

this extension is in line with the
link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
calling `submit_with_event` and the `handler` overloads of these enqueue
functions.

[source,c++]
----
namespace ext::oneapi::experimental {

> void *async_malloc(const queue &que, usm::alloc kind, size_t size);

Author: I thought about this as well, and we had an offline conversation about it and concluded that we should follow the convention in the spec and pass SYCL classes with reference semantics like `queue` by const ref, the motivation being partly that this is what the SYCL spec is currently doing, but also to follow C++ guidelines to use const ref for anything which isn't cheep to copy like a fundamental type, as even when implementing these types with a shared_ptr, this can be costly to copy.

> void *async_malloc(const queue &que, usm::alloc kind, size_t size);

Reviewer 7: I agree that we should have a discussion about whether it's best to pass `queue` by reference or value.  However, it makes no sense for this extension to be different from the others.  The style of this extension is supposed to be similar to  [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc) and [sycl_khr_free_function_commands](https://github.com/KhronosGroup/SYCL-Docs/pull/644), and both of those extensions pass `queue` by value.

Part of the problem, I think, is that this function conceptually changes the queue (by enqueuing a command to it), so it's not clear that `const` is the right thing.  However, passing the parameter as `queue &` (without the `const`) doesn't work because then you can't pass a temporary `queue` object as a parameter.

If we want to change the style to pass these parameters as `const &`, then we should have a broader discussion about the spec.  We should not just change this one extension.

> void *async_malloc(const queue &que, usm::alloc kind, size_t size);

Reviewer 7: I agree.  As you note, we have since discussed this more globally, and we have decided to pass `queue` (and other common reference semantic objects) by reference.  The KHR [sycl_khr_free_function_commands](https://github.com/KhronosGroup/SYCL-Docs/pull/644) has already been updated, and we think the other APIs should be updated too.

void *async_malloc(handler &cgh, usm::alloc kind, size_t size);

void *async_malloc_from_pool(const queue &que, size_t size,
  const memory_pool &pool);

void *async_malloc_from_pool(handler &cgh, size_t size,
  const memory_pool &pool);

void async_free(const queue &que, void *ptr);

void async_free(handler &cgh, void *ptr);

}  // ext::oneapi::experimental
----

[source, c++]
----
void *async_malloc(const queue &que, usm::alloc kind, size_t size);

void *async_malloc(handler &cgh, usm::alloc kind, size_t size);

void *async_malloc_from_pool(const queue &que, size_t size,
  const smemory_pool &pool);

void *async_malloc_from_pool(handler &cgh, size_t size,
  const memory_pool &pool);
----

> _Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which

Reviewer 11: ```suggestion
_Effects_: Enqueues a command to `que` or the SYCL queue associated with `cgh` which
```

Reviewer 11: There is a few other cases in this description where the variable names are incorrect.

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/d154a67861e46477ba07b415b845348f880ce8a4.

will asynchronously allocate memory of size `size` in bytes, allocating from the

> memory pool `pool` if provided, otherwise allocation from the default memory

Reviewer 11: ```suggestion
memory pool `pool` if provided, otherwise allocating from the default memory
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/d154a67861e46477ba07b415b845348f880ce8a4.

memory pool associated with the SYCL context and device associated with `que` or
`cgh`. If a memory pool `pool` is provided the allocation kind is that of the
memory pool, otherwise if an allocation kind `kind` parameter is provided, this
is the allocation kind which is used. If a memory pool `pool` is provided, this

> must be associated with the same SYCL context and device as `que` or the SYCL

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

queue associated with `cgh`. Memory can either be allocated directly from the
memory pool or allocated to the memory pool to provide enough memory in the

> memory pool for the allocation. Accessing the memory at the address of the

Author: Removed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

pointer returned by asynchronous malloc functions before the command has
completed execution is undefined behavior.

_Returns_: A pointer to the address of a memory reservation. When the allocation
size is zero bytes, these functions behave in a manner consistent with C++
`std::malloc`. The value returned is unspecified in this case, and the returned
pointer may not be used to access storage. If this pointer is not `null`, it
must be passed to `sycl::async_free` to avoid a memory leak.

_Throws_: An exception with the `errc::memory_allocation` error code if the
allocation brings the memory pool over the maximum size. This error must be
thrown asynchronously. An exception with the `errc::invalid` error code if

> `kind` is `usm::alloc::unknown`.

Author: Good idea, fixed in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

[source, c++]
----
void async_free(const queue &que, void *ptr);

void async_free(handler &cgh, void *ptr);
----

> _Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which

Reviewer 11: ```suggestion
_Effects_: Enqueues a command to `que` or the SYCL queue associated with `cgh` which
```

Author: Fixed in https://github.com/intel/llvm/pull/14800/commits/d154a67861e46477ba07b415b845348f880ce8a4.

which will asynchronously free the memory allocation at the address of `ptr`.
Accessing the memory at the address of `ptr` after the asynchronous free command
has completed execution is undefined behavior. If `ptr` is not the address of a

> memory allocation allocated to a memory pool this is undefined behavior. 

Author: The memory pool is associated with a context and potentially a device, so I would say that it should be allowed, though we should confirm this can be supported by the backends, so I agree for now, it would be safer to say it is not supported. I've applied this change in https://github.com/intel/llvm/pull/14800/commits/12ce607c720b79d85d11ad628f01320b245ba2a1.

  q.wait();
}
----

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
table below.  Applications can test for the existence of this macro to determine
if the implementation supports this feature, or applications can test the
macro's value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Querying async memory alloc support

The following device query is provided to query whether a SYCL implementation
provides support for async memory allocation on a particular device.

The device aspects for this query is:

[frame="none",options="header"]
|======================
|Device descriptor |Description
|`aspect::ext_oneapi_async_memory_alloc` | Indicates if the device supports 
asynchronous memory allocation and memory pools.

> |======================

Reviewer 7: Sorry, I thought I made this comment in the last round, but I can't find it now.

If we want an aspect like this, we also need to specify the behavior of the APIs when they are called with a device that does *not* have the aspect.  Usually, we throw `errc::feature_not_supported` in this case.  Offhand, I think this will affect the following APIs:

* The `memory_pool` constructors.
* The `ext_oneapi_get_default_memory_pool` functions.
* The `async_malloc` functions (which do not take a memory pool parameter).

There is an interesting question here.  What is the behavior of the `memory_pool` constructor that creates a *host* USM memory pool?  Does this fail if there is even one device in the context does not have the aspect?  Or does it fail only if no devices in the context have the aspect?  My first thought is the former, it fails unless *all* devices in the context have the aspect.

There is a similar issue with the form of `ext_oneapi_get_default_memory_pool` which return the default host memory pool.

Author: Good catch, I've added wording to describe the behaviour of APIs when the aspect is not supported in https://github.com/intel/llvm/pull/14800/commits/2deb534e1c57aa284c3cfdfdbc5509d7ba414c76.

> There is an interesting question here. What is the behavior of the memory_pool constructor that creates a host USM memory pool? Does this fail if there is even one device in the context does not have the aspect? Or does it fail only if no devices in the context have the aspect? My first thought is the former, it fails unless all devices in the context have the aspect.

So the way I was expecting this would work would be the former as well, since host USM allocations should be accessible to all devices, which I interpret to be all devices in the context, then it would make sense that all devices support async memory allocations. Though I suspect it's unlikely that one device on a platform will support it while another doesn't.
